File: app/src/main/java/com/example/scanner/ui/main/MainScreen.kt
package com.example.scanner.ui.main

import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.navigation3.runtime.NavKey
import com.example.scanner.Scanner
import com.example.scanner.data.DefaultDataRepository
import com.example.scanner.theme.ScannerTheme

@Composable
fun MainScreen(
  onItemClick: (NavKey) -> Unit,
  modifier: Modifier = Modifier,
  viewModel: MainScreenViewModel = viewModel { MainScreenViewModel(DefaultDataRepository()) },
) {
  val state by viewModel.uiState.collectAsStateWithLifecycle()
  when (state) {
    MainScreenUiState.Loading -> {
      // Blank
    }
    is MainScreenUiState.Success -> {
      MainScreen(
        data = (state as MainScreenUiState.Success).data,
        onScanClick = { onItemClick(Scanner) },
        modifier = modifier
      )
    }
    is MainScreenUiState.Error -> {
      Text("Error loading data: ${(state as MainScreenUiState.Error).throwable.message}")
    }
  }
}

@Composable
internal fun MainScreen(
  data: List<String>,
  onScanClick: () -> Unit,
  modifier: Modifier = Modifier
) {
  Column(
    modifier = modifier.fillMaxSize(),
    horizontalAlignment = Alignment.CenterHorizontally,
    verticalArrangement = Arrangement.Center
  ) {
    Text(
      text = "Barcode Scanner",
      style = MaterialTheme.typography.headlineLarge,
      textAlign = TextAlign.Center
    )

    Spacer(modifier = Modifier.height(16.dp))

    Text(
      text = "Scan barcodes and QR codes instantly",
      style = MaterialTheme.typography.bodyLarge,
      textAlign = TextAlign.Center,
      color = MaterialTheme.colorScheme.onSurfaceVariant
    )

    Spacer(modifier = Modifier.height(48.dp))

    Button(
      onClick = onScanClick,
      modifier = Modifier
        .fillMaxWidth()
        .height(56.dp),
      shape = MaterialTheme.shapes.large
    ) {
      Icon(
        imageVector = Icons.Default.QrCodeScanner,
        contentDescription = null,
        modifier = Modifier.size(24.dp)
      )
      Spacer(modifier = Modifier.width(12.dp))
      Text(
        text = "Start Scanning",
        style = MaterialTheme.typography.titleMedium
      )
    }

    Spacer(modifier = Modifier.height(24.dp))

    data.forEach { Greeting(it) }
  }
}

@Composable
fun Greeting(name: String, modifier: Modifier = Modifier) {
  Text(text = "Hello $name!", modifier = modifier)
}

@Preview(showBackground = true)
@Composable
fun MainScreenPreview() {
  ScannerTheme { MainScreen(listOf("Android"), onScanClick = {}) }
}

@Preview(showBackground = true, widthDp = 340)
@Composable
fun MainScreenPortraitPreview() {
  ScannerTheme { MainScreen(listOf("Android"), onScanClick = {}) }
}

File: ios/Scanner/UI/Main/MainScreen.swift
import SwiftUI

/// Equivalent of the Android MainScreen.
struct MainScreen: View {
    let onScan: () -> Void

    @StateObject private var viewModel: MainScreenViewModel

    init(
        onScan: @escaping () -> Void,
        viewModel: MainScreenViewModel = MainScreenViewModel(repository: DefaultDataRepository())
    ) {
        self.onScan = onScan
        _viewModel = StateObject(wrappedValue: viewModel)
    }

    var body: some View {
        VStack(spacing: 0) {
            Spacer()

            Text("Barcode Scanner")
                .font(ScannerTheme.Typography.headline)
                .multilineTextAlignment(.center)

            Spacer().frame(height: 16)

            Text("Scan barcodes and QR codes instantly")
                .font(ScannerTheme.Typography.body)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)

            Spacer().frame(height: 48)

            Button(action: onScan) {
                Label("Start Scanning", systemImage: "qrcode.viewfinder")
                    .font(.title3)
                    .frame(maxWidth: .infinity)
                    .frame(height: 56)
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)

            Spacer().frame(height: 24)

            switch viewModel.uiState {
            case .loading:
                EmptyView()
            case .error(let message):
                Text("Error loading data: \(message)")
                    .foregroundStyle(.red)
            case .success(let data):
                ForEach(data, id: \.self) { item in
                    Text("Hello \(item)!")
                }
            }

            Spacer()
        }
        .padding(16)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

#Preview {
    MainScreen(onScan: {})
}
